### ISSUE TITLE
Set/Map/Multimap: Custom equality?

### ISSUE THREAD

**Issue by matanlurey** [google/built_collection.dart#76]
Any concern about support `package:collection`'s `Equality` class, optionally?

I have some use cases where I'd like to use BuiltCollection(s), but the keys I don't have control over and don't really want to write wrapper objects just to change the identity.

If you're OK, I can send PRs. :)

**Comment by davidmorgan** [comment]
Sounds reasonable. I guess the cost is an extra field to optionally store the equality? We could presumably optimize that by having a different concrete implementation, but I don't think that's necessary.

It'll have to pass through the builders, too, so when you rebuild you keep the same equality by default. It might make sense to expose the equality on the builder, so you can rebuild and change it:

final myList = new BuiltList<int>([1, 2, 3], equality: new EverythingIsEqual());
final mySecondList = myList.rebuild((b) => b.equality = new NothingIsEqual());

PRs sound good. Thanks!

**Comment by matanlurey** [comment]
@davidmorgan How would you like to handle this API change:

(Current)

```dart
factory BuiltSet([Iterable iterable = const []]) { ... }
```

Option 1: Add a new named constructor (No breaking change)

```dart
factory BuiltSet.withEquality(Equality equality, [Iterable iterable = const []])
```

Option 2: Modify the default constructor (Breaking change)

```dart
factory BuiltSet({Iterable from: const [], Equality equality})
```

Option 3: Modify the default constructor (No breaking change, but awkward)

```dart
factory BuiltSet([Iterable iterable = const [], Equality equality])
```

**Comment by matanlurey** [comment]
On another hand, which actually helps cover https://github.com/google/built_collection.dart/issues/73, maybe we can just support a custom factory method for a given builder, i.e.:

```dart
factory BuiltSet.createWith(Set<E> newSet(Iterable<E> iterable), [Iterable iterable = const []);
```

Then you can support whatever you want:

```dart
new BuiltSet.createWith((iterable) => MySpecialSet<E>(iterable));
````

**Comment by davidmorgan** [comment]
Hmm, interesting, but I don't see how it would help with equality -- we don't use anything from the underlying set for equality?

Another API option would be to only set equality via the builder. Then you can use the existing "build" factory:

new BuiltSet<int>.build((b) => b..replace([1, 2, 3])..equality = new MyEquality());

It's not beautiful, but I think we would want to support this anyway, so it defers the question of what to do with the BuiltSet factories.

One note: should it be called elementEquality? Then it's clear it's on elements. I'm not sure if we want to mess with the collection equality?

**Comment by matanlurey** [comment]
It would be help w/ equality because you could create a set with custom equality:

```dart
new BuiltSet.createWith((iterable) => new HashSet(hashCode: (_) => 1)..addAll(iterable));
```

**Comment by davidmorgan** [comment]
Got it. That is pretty neat. Please do send PRs.

I guess we can say something along the lines of "Please note that BuiltSet uses a LinkedHashSet by default, and that changing the underlying set implementation can lose some guarantees. For example, using a HashSet means you lose ordering guarantees."

Not sure about "create" in the name when we already have "new". How about: custom, withFactory, withSet, withSetFactory, withCustomSet, ... ?

**Comment by pschiffmann** [comment]
I started working on this [here](https://github.com/pschiffmann/built_collection.dart/pull/1) and would like to hear your opinions on this before I continue with the other classes, especially on these questions:

1. Should `base`/`_setConstructor` be exchangeable later on, or should it be fixed after instantiation?

2. If `base` can be changed, should it be changed when I call `builder.replace(s)`, and `s` uses another base?

3. Will two sets with different base sets and equal elements compare as equal?

4. @matanlurey Why did you include an `iterable` parameter in your example?

Further ideas for the constructor name: `.using()`, `.basedOn()`, `.wrapping()`, `.of()`, `.over()`.

**Comment by davidmorgan** [comment]
Thanks Philipp! The code looks good so far.

Re: how to design the API; I still quite like the builder-based approach in my update from Nov 10 2016. this means we don't need a new constructor, but instead you would do

```
new BuiltSet<int>.build((b) => b
    ..base = () => new SplayTreeSet()
    ..addAll([...]));
```

we could then defer whether to add new constructors. What do you think?

For dart 2 we're going to need to add new constructors, to differentiate new BuiltSet(Iterable<E>) from new BuiltSet.from(Iterable). This will be a breaking change. So we could potentially do more breaking changes at the same time if that helps.

To your specific questions:

1. I think fixed makes sense, unless we decide that using the builder to set it is a better API, in which case it will need to be settable in the builder :)
2. I think it should be only changed explicitly.
3. That's tough. I don't have a strong opinion from a design point of view. From an _implementation_ point of view, I don't think there's a good way to compare/hash based on the `createSet` function ... so the path of least resistance is to leave operator== as it is.
4. Have to let Matan answer that one :) ... I guess the flexibility it adds is that you could conceivably return a different implementation based on the properties of the iterable; I don't really see that being a big benefit.

**Comment by matanlurey** [comment]
I don't 100% recall :)

**Comment by pschiffmann** [comment]
> ```
> new BuiltSet<int>.build((b) => b
>     ..base = () => new SplayTreeSet()
>     ..addAll([...]));
> ```

I considered that version too, but a setter without getter felt awkward, and adding one would be the first time a builder exposed any of it's state. And because your code didn't provide any blueprints/inspiration on that matter, I refrained from it ;)
Also, it introduces another problem: I currently use `null` to indicate _default base_ (because the lambda expression `() => new Set<E>()` isn't equal with itself, which would fail [this](https://github.com/pschiffmann/built_collection.dart/pull/1/files#diff-e90e3c2d618e9e0f95df21c98d546422R71) condition). But I think it looks unintuitive from a user perspective to write `builder.base = null` to reset base to the default value. Another option would be
```
class SetBuilder {
  static Set defaultBase() => new Set();
}
```
and use that by default, if we are willing to give up the generic type.
Or we could just remove the reuse feature from the `replace` method.
So if you can live with one of these (or have a better idea) I'd favor the setter over the constructor, too. :)

> For dart 2 we're going to need to add new constructors, to differentiate new BuiltSet(Iterable) from new BuiltSet.from(Iterable).

Is that a typo?

> That's tough. I don't have a strong opinion from a design point of view. From an implementation point of view, I don't think there's a good way to compare/hash based on the createSet function ... so the path of least resistance is to leave operator== as it is.

Sounds good to me.

**Comment by davidmorgan** [comment]
Sorry for the slow response; have been feeling unwell the last few days.

It's fine for the builder to have methods if we want the property to be write-only:

```
b.withBase(() => ...)
```

and with that, it seems reasonable to have something like  `withDefaultBase` to reset it:

```
..b.withDefaultBase()
```

It was a typo, yes -- Github swallowed the `<>` in one of them. Now fixed :)

**Comment by pschiffmann** [comment]
Take your time and recover – I'm just a bored student with too much time on my hands, I can wait a few days for a response :-)

I'll update my branch and file a PR on your repository.

**Comment by davidmorgan** [comment]
Where would open source be without bored students ;)

Sounds good to me, thanks.

**Comment by pschiffmann** [comment]
Opened #115.

**Comment by pschiffmann** [comment]
Do you have a good idea what I can do with the multimaps? The builders need two different maps, `Map<K, BuiltList<V>> _builtMap` and `Map<K, ListBuilder<V>> _builderMap`. The common type that `base` could return is `Map<K, Iterable<V>>`, but then every method would have to cast to the actual type.

And regarding the set multimap, which do you prefer:
```
class SetMultimapBuilder<K, V> {
  void withMapBase(Map<K, Iterable<V> Function() base);
  void withDefaultMapBase();
  /// If you need a custom set type or equality for the sets,
  /// set the base of the builder that is returned from `base` accordingly.
  void withSetBase(SetBuilder<V> Function() base);
  void withDefaultSetBase();
}
```
or
```
class SetMultimapBuilder<K, V> {
  void withBase({
    Map<K, Iterable<V> Function() mapBase,
    SetBuilder<V> Function() setBase});
  void withDefaultBase({bool map: true, bool set: true});
}
```
or something else?

**Comment by davidmorgan** [comment]
Fun fun.

`ListBuilder` is actually _not_ an `Iterable` right now, so actually it would have to be `Map<K, Object>`, I think? An alternative would be to accept a `Function<K, V>`.

I'd be tempted to wait before adding to the multimaps--unless you have a need for it yourself right now. Gives a bit of time to see how the set/map are received. What do you think?

**Comment by pschiffmann** [comment]
Sure, I personally don't need multimaps right now, I just looked at them for the sake of completeness.

You want to release a new version, then? Or is there anything else left to do on this topic right now?

**Comment by davidmorgan** [comment]
Sounds good. Yes, I'll publish to pub on Monday. Thanks!

**Comment by davidmorgan** [comment]
This is now released as v2.1.0.
